### PR TITLE
edge_id op csr forward on CPU

### PR DESCRIPTION
**Pull request by HyperZealot**
## Description ##
You know what this is...

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] mx.nd.contrib.edge_id(data, u, v)
- [x] corresponding unit tests

## Comments ##
Passed >10000 times for flakiness check

### CODE REVIEW

**Comment by HyperZealot** [comment]
@zheng-da

File: src/operator/contrib/edge_id-inl.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                        std::vector<TShape>* in_attrs,
+                        std::vector<TShape>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  CHECK_EQ(in_attrs->at(1).ndim(), 1U);
+  CHECK_EQ(in_attrs->at(2).ndim(), 1U);
+  CHECK_EQ(in_attrs->at(1)[0], in_attrs->at(2)[0]);
+
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(1));
+  SHAPE_ASSIGN_CHECK(*in_attrs, 1, out_attrs->at(0));
```
Comment (zheng-da) [inline src/operator/contrib/edge_id-inl.h:47]: shouldn't you infer in_attrs[2] with out_attrs as well?

File: src/operator/contrib/edge_id-inl.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  }
+  return dispatched;
+}
+
+struct edge_id_csr_forward {
+  template<typename DType, typename IType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const DType* in_data,
+                                  const IType* in_indices, const IType* in_indptr,
+                                  const CType* u, const CType* v) {
+    const int target_row_id = static_cast<int>(u[i]);
```
Comment (zheng-da) [inline src/operator/contrib/edge_id-inl.h]: why do you cast it to int?

File: src/operator/contrib/edge_id-inl.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    const int target_col_id = static_cast<int>(v[i]);
+    out_data[i] = DType(-1);
+    for (IType idx = in_indptr[target_row_id];
+         idx < in_indptr[target_row_id + 1];
+         ++idx) {
+      if (in_indices[idx] == target_col_id) {
+        out_data[i] = in_data[idx];
+      }
+    }
```
Comment (zheng-da) [inline src/operator/contrib/edge_id-inl.h:97]: you can do binary search. There is an implementation in C++ STD.

File: src/operator/contrib/edge_id-inl.h (hunk 1 above)
Comment (HyperZealot) [reply src/operator/contrib/edge_id-inl.h:47]: okay

File: src/operator/contrib/edge_id-inl.h (hunk 2 above)
Comment (HyperZealot) [reply src/operator/contrib/edge_id-inl.h]: so that u & v could be of any data type?

File: src/operator/contrib/edge_id-inl.h (hunk 3 above)
Comment (HyperZealot) [reply src/operator/contrib/edge_id-inl.h:97]: okay

**Pull request closed without merging**